### ISSUE TITLE
Use interpolated strings instead of String.Concat or String.Join

### ISSUE THREAD

**Issue by wmjordan** [dotnet/roslyn#58557]
**Brief description:**

I've gotten some legacy code which is rewritten about 10 years ago where a lot of `String.Concat(string[])` or `String.Join(string, params string[])` are used. Can you provide a code refactoring to change them to interpolated strings?

**Languages applicable:**

C# (maybe VB.NET also)

**Code example that the analyzer should report:**

````C#
String.Concat("Hello ", user.Name, "! Time is ", DateTime.Now.ToString());
logger.Log(String.Join(" ", DateTime.Now.ToString(), LogLevel.Info.ToString(), logCategory, logContent));
````

**Comment by CyrusNajmabadi** [comment]
I think this would def be a nice to have.  A cleanly implemented feature that implemented this would likely be accepted.  Open question if this should be a refactoring or an analyzer/fixer combo.

**Comment by Youssef1313** [comment]
> Open question if this should be a refactoring or an analyzer/fixer combo.

In cases where this will benefit from interpolated string handlers and can improve performance, I think it should be an analyzer diagnostic.

Otherwise, a refactoring is likely to be sufficient. Does that sound reasonable?

**Comment by ameisen** [comment]
I should note that as of 12/8's build of Roslyn (the most recent on SharpLab) generates... very bad IL for interpolated strings, and thus generates _awful_ JIT output.

```csharp
public static class C {    
    public static unsafe string M0(string a, string b) {
        return "{{" + a + "||" + b + "}}";
    }
    
    public static unsafe string M1(string a, string b) {
        return $"{{{{{a}||{b}}}}}";
    }
}
```

IL:

```msil
.method public hidebysig static 
		string M0 (
				string a,
				string b
		) cil managed 
{
		// Method begins at RVA 0x2050
		// Code size 44 (0x2c)
		.maxstack 8

		IL_0000: ldc.i4.5
		IL_0001: newarr [System.Private.CoreLib]System.String
		IL_0006: dup
		IL_0007: ldc.i4.0
		IL_0008: ldstr "{{"
		IL_000d: stelem.ref
		IL_000e: dup
		IL_000f: ldc.i4.1
		IL_0010: ldarg.0
		IL_0011: stelem.ref
		IL_0012: dup
		IL_0013: ldc.i4.2
		IL_0014: ldstr "||"
		IL_0019: stelem.ref
		IL_001a: dup
		IL_001b: ldc.i4.3
		IL_001c: ldarg.1
		IL_001d: stelem.ref
		IL_001e: dup
		IL_001f: ldc.i4.4
		IL_0020: ldstr "}}"
		IL_0025: stelem.ref
		IL_0026: call string [System.Private.CoreLib]System.String::Concat(string[])
		IL_002b: ret
} // end of method C::M0

.method public hidebysig static 
		string M1 (
				string a,
				string b
		) cil managed 
{
		// Method begins at RVA 0x2080
		// Code size 69 (0x45)
		.maxstack 3
		.locals init (
				[0] valuetype [System.Private.CoreLib]System.Runtime.CompilerServices.DefaultInterpolatedStringHandler
		)

		IL_0000: ldloca.s 0
		IL_0002: ldc.i4.6
		IL_0003: ldc.i4.2
		IL_0004: call instance void [System.Private.CoreLib]System.Runtime.CompilerServices.DefaultInterpolatedStringHandler::.ctor(int32, int32)
		IL_0009: ldloca.s 0
		IL_000b: ldstr "{{"
		IL_0010: call instance void [System.Private.CoreLib]System.Runtime.CompilerServices.DefaultInterpolatedStringHandler::AppendLiteral(string)
		IL_0015: ldloca.s 0
		IL_0017: ldarg.0
		IL_0018: call instance void [System.Private.CoreLib]System.Runtime.CompilerServices.DefaultInterpolatedStringHandler::AppendFormatted(string)
		IL_001d: ldloca.s 0
		IL_001f: ldstr "||"
		IL_0024: call instance void [System.Private.CoreLib]System.Runtime.CompilerServices.DefaultInterpolatedStringHandler::AppendLiteral(string)
		IL_0029: ldloca.s 0
		IL_002b: ldarg.1
		IL_002c: call instance void [System.Private.CoreLib]System.Runtime.CompilerServices.DefaultInterpolatedStringHandler::AppendFormatted(string)
		IL_0031: ldloca.s 0
		IL_0033: ldstr "}}"
		IL_0038: call instance void [System.Private.CoreLib]System.Runtime.CompilerServices.DefaultInterpolatedStringHandler::AppendLiteral(string)
		IL_003d: ldloca.s 0
		IL_003f: call instance string [System.Private.CoreLib]System.Runtime.CompilerServices.DefaultInterpolatedStringHandler::ToStringAndClear()
		IL_0044: ret
} // end of method C::M1
```

JIT output:

```assembly
C.M0(System.String, System.String)
	L0000: push edi
	L0001: push esi
	L0002: mov esi, ecx
	L0004: mov edi, edx
	L0006: mov ecx, 0x6ceedf0
	L000b: mov edx, 5
	L0010: call 0x063e31f8
	L0015: lea edx, [eax+8]
	L0018: mov ecx, [0x8d95b64]
	L001e: call 0x7263b7a0
	L0023: lea edx, [eax+0xc]
	L0026: call 0x7263b7d0
	L002b: lea edx, [eax+0x10]
	L002e: mov ecx, [0x8d95344]
	L0034: call 0x7263b7a0
	L0039: lea edx, [eax+0x14]
	L003c: call 0x7263b800
	L0041: lea edx, [eax+0x18]
	L0044: mov ecx, [0x8d95b60]
	L004a: call 0x7263b7a0
	L004f: mov ecx, eax
	L0051: call System.String.Concat(System.String[])
	L0056: pop esi
	L0057: pop edi
	L0058: ret

C.M1(System.String, System.String)
	L0000: push ebp
	L0001: mov ebp, esp
	L0003: push edi
	L0004: push esi
	L0005: push ebx
	L0006: sub esp, 0x1c
	L0009: vxorps xmm4, xmm4, xmm4
	L000d: vmovdqu [ebp-0x24], xmm4
	L0012: xor eax, eax
	L0014: mov [ebp-0x14], eax
	L0017: mov [ebp-0x10], eax
	L001a: mov esi, ecx
	L001c: mov edi, edx
	L001e: mov ecx, [0x8d72000]
	L0024: mov edx, 0x100
	L0029: cmp [ecx], ecx
	L002b: call dword ptr [0xb4d7bf4]
	L0031: lea ebx, [ebp-0x24]
	L0034: mov [ebp-0x20], eax
	L0037: test eax, eax
	L0039: je L01bd
	L003f: lea edx, [eax+8]
	L0042: mov eax, [eax+4]
	L0045: add ebx, 0x10
	L0048: mov [ebx], edx
	L004a: mov [ebx+4], eax
	L004d: xor ecx, ecx
	L004f: mov [ebp-0x1c], ecx
	L0052: mov byte ptr [ebp-0x18], 0
	L0056: lea ecx, [ebp-0x14]
	L0059: mov edx, [ecx]
	L005b: mov [ebp-0x28], edx
	L005e: mov ecx, [ecx+4]
	L0061: mov eax, [ebp-0x1c]
	L0064: xor ebx, ebx
	L0066: dec ecx
	L0067: mov edx, ecx
	L0069: sar edx, 0x1f
	L006c: cmp eax, ecx
	L006e: sbb ebx, edx
	L0070: jge L01c6
	L0076: mov edx, [ebp-0x28]
	L0079: lea ecx, [edx+eax*2]
	L007c: mov edx, [0x8d95b64]
	L0082: add edx, 8
	L0085: mov edx, [edx]
	L0087: mov [ecx], edx
	L0089: add eax, 2
	L008c: mov [ebp-0x1c], eax
	L008f: cmp byte ptr [ebp-0x18], 0
	L0093: jne L014d
	L0099: test esi, esi
	L009b: je L014d
	L00a1: mov ecx, [ebp-0x1c]
	L00a4: cmp ecx, [ebp-0x10]
	L00a7: ja L01d9
	L00ad: mov edx, [ebp-0x14]
	L00b0: mov eax, [ebp-0x10]
	L00b3: sub eax, ecx
	L00b5: lea ecx, [edx+ecx*2]
	L00b8: mov edx, [esi+4]
	L00bb: cmp edx, eax
	L00bd: ja L014d
	L00c3: lea edx, [esi+8]
	L00c6: mov eax, [esi+4]
	L00c9: add eax, eax
	L00cb: push eax
	L00cc: call System.Buffer.Memmove(Byte ByRef, Byte ByRef, UIntPtr)
	L00d1: mov ecx, [ebp-0x1c]
	L00d4: add ecx, [esi+4]
	L00d7: mov [ebp-0x1c], ecx
	L00da: lea ecx, [ebp-0x14]
	L00dd: mov edx, [ecx]
	L00df: mov ecx, [ecx+4]
	L00e2: mov eax, [ebp-0x1c]
	L00e5: xor esi, esi
	L00e7: dec ecx
	L00e8: mov ebx, ecx
	L00ea: sar ebx, 0x1f
	L00ed: cmp eax, ecx
	L00ef: sbb esi, ebx
	L00f1: jge L01df
	L00f7: lea ecx, [edx+eax*2]
	L00fa: mov edx, [0x8d95344]
	L0100: add edx, 8
	L0103: mov edx, [edx]
	L0105: mov [ecx], edx
	L0107: add eax, 2
	L010a: mov [ebp-0x1c], eax
	L010d: cmp byte ptr [ebp-0x18], 0
	L0111: jne L01a1
	L0117: test edi, edi
	L0119: je L01a1
	L011f: mov ecx, [ebp-0x1c]
	L0122: cmp ecx, [ebp-0x10]
	L0125: ja L01d9
	L012b: mov edx, [ebp-0x14]
	L012e: mov eax, [ebp-0x10]
	L0131: sub eax, ecx
	L0133: lea ecx, [edx+ecx*2]
	L0136: mov edx, [edi+4]
	L0139: cmp edx, eax
	L013b: ja short L01a1
	L013d: lea edx, [edi+8]
	L0140: mov eax, [edi+4]
	L0143: add eax, eax
	L0145: push eax
	L0146: call System.Buffer.Memmove(Byte ByRef, Byte ByRef, UIntPtr)
	L014b: jmp short L0159
	L014d: lea ecx, [ebp-0x24]
	L0150: mov edx, esi
	L0152: call System.Runtime.CompilerServices.DefaultInterpolatedStringHandler.AppendFormattedSlow(System.String)
	L0157: jmp short L00da
	L0159: mov ecx, [ebp-0x1c]
	L015c: add ecx, [edi+4]
	L015f: mov [ebp-0x1c], ecx
	L0162: lea ecx, [ebp-0x14]
	L0165: mov eax, [ecx]
	L0167: mov ecx, [ecx+4]
	L016a: mov edx, [ebp-0x1c]
	L016d: xor edi, edi
	L016f: dec ecx
	L0170: mov esi, ecx
	L0172: sar esi, 0x1f
	L0175: cmp edx, ecx
	L0177: sbb edi, esi
	L0179: jge short L01ad
	L017b: lea ecx, [eax+edx*2]
	L017e: mov eax, [0x8d95b60]
	L0184: add eax, 8
	L0187: mov eax, [eax]
	L0189: mov [ecx], eax
	L018b: add edx, 2
	L018e: mov [ebp-0x1c], edx
	L0191: lea ecx, [ebp-0x24]
	L0194: call System.Runtime.CompilerServices.DefaultInterpolatedStringHandler.ToStringAndClear()
	L0199: lea esp, [ebp-0xc]
	L019c: pop ebx
	L019d: pop esi
	L019e: pop edi
	L019f: pop ebp
	L01a0: ret
	L01a1: lea ecx, [ebp-0x24]
	L01a4: mov edx, edi
	L01a6: call System.Runtime.CompilerServices.DefaultInterpolatedStringHandler.AppendFormattedSlow(System.String)
	L01ab: jmp short L0162
	L01ad: lea ecx, [ebp-0x24]
	L01b0: mov edx, [0x8d95b60]
	L01b6: call System.Runtime.CompilerServices.DefaultInterpolatedStringHandler.GrowThenCopyString(System.String)
	L01bb: jmp short L0191
	L01bd: xor edx, edx
	L01bf: xor eax, eax
	L01c1: jmp L0045
	L01c6: lea ecx, [ebp-0x24]
	L01c9: mov edx, [0x8d95b64]
	L01cf: call System.Runtime.CompilerServices.DefaultInterpolatedStringHandler.GrowThenCopyString(System.String)
	L01d4: jmp L008f
	L01d9: call System.ThrowHelper.ThrowArgumentOutOfRangeException()
	L01de: int3
	L01df: lea ecx, [ebp-0x24]
	L01e2: mov edx, [0x8d95344]
	L01e8: call System.Runtime.CompilerServices.DefaultInterpolatedStringHandler.GrowThenCopyString(System.String)
	L01ed: jmp L010d
```

**Comment by CyrusNajmabadi** [comment]
> I should note that as of 12/8's build of Roslyn (the most recent on SharpLab) generates... very bad IL for interpolated strings

Can you clarify what is bad about that IL?

**Comment by ameisen** [comment]
> > I should note that as of 12/8's build of Roslyn (the most recent on SharpLab) generates... very bad IL for interpolated strings
> 
> Can you clarify what is bad about that IL?

It's completely unoptimized. On other builds they both generate the _same_ IL.

**Comment by CyrusNajmabadi** [comment]
> It's completely unoptimized. 

What metric are you using to evaluate that?

> The JIT result for it results in output that is 6x longer, 

Length isn't relevant here. Amount of work and allocations are. 

> does way more work

What are you defining as work here?

For example the old code requires a full allocation of an array to do the concat call.  The new code requires no unnecessary intermediary allocation.

**Comment by CyrusNajmabadi** [comment]
Tagging @333fred to provide more information.

**Comment by CyrusNajmabadi** [comment]
Note: the il we generate looks exactly to be what I would expect. The JIT really looks to be the result of heavy inlining. 

If the JIT result is not desirable for you, I think you'd have to take this up with dotnet/runtime for them to weigh in on this aspect. 

Note: my guess is that if string.Concat got inlined here, you'd see similar complexity in jit output.

**Comment by 333fred** [comment]
@ameisen I would suggest taking a read through @stephentoub's excellent blog post on the changes to interpolated strings in .NET 6 and C# 10, and the measurable performance improvements it brings: https://devblogs.microsoft.com/dotnet/string-interpolation-in-c-10-and-net-6/. In particular, his example of formatting a version string sees a 40% throughput increase and a nearly 5x reduction in allocations. The IL may be longer, but that does not mean that it is less efficient.

**Comment by wmjordan** [comment]
According to the discussion above, changing from `String.Concat(string[])` or `String.Join(string, string[])` to string interpolation seems to even help improve the performance in some cases (we don't need to create arrays for them any more). Perhaps an analyzer for them is also appropriated.

**Comment by CyrusNajmabadi** [comment]
hi @wmjordan, could you explain what you mean here by "Perhaps an analyzer for them is also appropriated."?  Thanks!

**Comment by wmjordan** [comment]
@CyrusNajmabadi 

I meant as @Youssef1313 said in the previous comment:

> In cases where this will benefit from interpolated string handlers and can improve performance, I think it should be an analyzer diagnostic.

**Comment by CyrusNajmabadi** [comment]
> In cases where this will benefit from interpolated string handlers and can improve performance, I think it should be an analyzer diagnostic.

I'm not really sure waht the analyzer would say.  The compiler will alredy pick the efficient overload if it is available.  :)

**Comment by Youssef1313** [comment]
@CyrusNajmabadi What I meant is the following:

- If the overload that makes use of a string interpolation handler exists, but user is using `string.Concat` or `string.Join`, then the analyzer would suggest using string interpolation.
- If no overload that makes use of a string interpolation handler exists, then just a refactoring to use string interpolation would be sufficient.

**Comment by CyrusNajmabadi** [comment]
I see.  Sure, that seems reasonable.
